fluid/conv_seq_to_seq/utils.py

```diff
@@ -0,0 +1,250 @@
+        assert len(dims) == 3, "format shoud be BTC, get shape: %s" % str(dims)
+        # conv2d's input should be format NCHW, which is
+        # - batch_size, channels, height, width
+        B, T, C = dims
+        x = Op.transpose(x, [1, 2])
+        x = Op.reshape(x, (B, C, 1, T))
+
+        # here something bug, conv2d will change the original width and height by padding.
+        # just use a trick to map to the original size
+        N, C, H, W = get_tensor(x).dims
+
+        #H = (H + 2 * padding - kernel_size) / stride + 1
+        if self.atom is None:
+            size = [H, W]
+            self.kernel_size = [1, self.kernel_size]
+            padding = [(
+                (size[i] - 1) * self.stride + self.kernel_size[i] - size[i]) / 2
+                       for i in range(2)]
+            self.atom = conv2d(
+                self.out_channels, self.kernel_size, padding=padding)
+
+        x = self.atom(x)
+
+        x = Op.reshape(x, (B, T, -1))
+        return x
+
+
+class Atom(object):
+    counter = 0
+
+    def __init__(self, op, prefix="atom", **kwargs):
+        self.op = op
+        assert 'param_attr' not in kwargs
+        self.kwargs = kwargs
+        self.name = '%s-%d' % (prefix, Atom.counter)
+        Atom.counter += 1
+
+    def __call__(self, x):
+        if 'dropout' in self.kwargs:
+            dropout = self.kwargs['dropout']
+            del self.kwargs['dropout']
+            if dropout is not None:
+                x = self.op(x,
+                            param_attr=fluid.ParamAttr(name=self.name),
+                            **self.kwargs)
+                return Op.dropout(x, dropout, is_test=is_test)
+
+        return self.op(x,
+                       param_attr=fluid.ParamAttr(name=self.name),
+                       **self.kwargs)
+
+
+def get_var_desc(var):
+    protostr = var.desc.serialize_to_string()
+    proto = framework_pb2.VarDesc.FromString(str(protostr))
+    return proto
+
+
+def get_tensor(var):
+    proto = get_var_desc(var)
+    return proto.type.lod_tensor.tensor
+
+
+def get_dims(var):
+    return get_tensor(var).dims
+
+
+def log(*args):
+    res = ""
+    for a in args:
+        res += " " + str(a)
+    logging.warning(res)
+
+
+class objdic(dict):
+    def __setattr__(self, key, value):
+        self[key] = value
+
+    def __getattr__(self, key):
+        return self[key]
+
+
+def prepare_data(sents, start_id, end_id, pad_id, max_len, offset=0):
+    res = []
+    for sent in sents:
+        # if offset == 1:
+        #     if sent[0] == start_id:
+        #         sent[0] = pad_id
+        #     sent.insert(0, start_id)
+        if offset == -1:
+            if sent[0] == start_id:
+                del sent[0]
+        else:
+            if sent[0] != start_id:
+                sent.insert(0, start_id)
+
+        if sent[-1] != end_id:
+            sent.append(end_id)
+        if len(sent) < max_len:
+            sent += [pad_id for i in xrange(max_len - len(sent))]
```

Comment:
Maybe we can pad the batch data according to the max length in current batch and allow variant max sequence length across batches, but this may also affect the model configurations such as the above ` x = Op.reshape(x, (B, T, -1))`